### PR TITLE
build: bump fbw dev env

### PR DESCRIPTION
**Pull request by tracernz**
Fixes #[issue_no]

## Summary of Changes

Bump FBW dev env to the latest with SDK 0.24.

## Additional context

Discord username (if different from GitHub):

## Testing instructions

Check that the aircraft generally functions okay in SU14 (non-beta).

## How to download the PR for QA

Every new commit to this PR will cause new A32NX and A380X artifacts to be created, built, and uploaded.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the left side, click on the bottom **PR** tab
1. Click on either **flybywire-aircraft-a320-neo** or **flybywire-aircraft-a380-842** download link at the bottom of the page

### CODE REVIEW

**Review by 2hwk — APPROVED** [review]

**Comment by Gurgel100** [comment]
Please also adjust the Rust version to 1.74 in rust-toolchain.toml

**Comment by frankkopp** [comment]
> Please also adjust the Rust version to 1.74 in rust-toolchain.toml

I assume you tested this? A while ago when I updated the dev env this lead to compile issues.

**Comment by Saschl** [comment]
> > Please also adjust the Rust version to 1.74 in rust-toolchain.toml
> 
> I assume you tested this? A while ago when I updated the dev env this lead to compile issues.

I did and everything worked fine 👍  In fact we could even go to 1.77 from what I could see. 1.74 is default on dev-env though. If we go to 1.77 without a new image, that would work but will cause download of rust dependencies on every build.

**Comment by 2hwk** [comment]
Note: This PR causes issues with builds running in SU14, do not merge until SU15 release.

**Comment by tracernz** [comment]
> Please also adjust the Rust version to 1.74 in rust-toolchain.toml

Bumped to the 1.74.1 patch release.

**Comment by Saschl** [comment]
> The dev-env only specifies 1.74 which is a bit loose for production IMO. Which exact release will it have in any given scenario?

What counts is the version in `rust-toolchain.toml`, it will override whatever is specified as default in the dev-env.

**Comment by tracernz** [comment]
> > The dev-env only specifies 1.74 which is a bit loose for production IMO. Which exact release will it have in any given scenario?
> 
> What counts is the version in `rust-toolchain.toml`, it will override whatever is specified as default in the dev-env.

Indeed, but better to make them match so a download isn’t required. It takes a while.

**Comment by alepouna** [comment]
SU15 has just been released, RTM?

**Comment by Saschl** [comment]
There’s one sdk update we missed, might want to incorporate that so we only have the change once 👍

File: fbw-a32nx/src/wasm/fbw_a320/build.sh [hunk 1]
```diff
@@ -79,6 +79,12 @@ clang++ \
   -fno-exceptions \
   -fms-extensions \
   -fvisibility=hidden \
+  -fdata-sections \
+  -fno-stack-protector \
+  -fstack-size-section \
+  -fwritable-strings \
```
Comment (tracernz) [inline fbw-a32nx/src/wasm/fbw_a320/build.sh:85]:
```suggestion
```
I don't think this flag is a good idea. It's to allow old code that expected to be able to write to string constants.

File: fbw-a32nx/src/wasm/fbw_a320/build.sh [hunk 2]
```diff
@@ -13,7 +13,7 @@ if [ "$1" == "--debug" ]; then
   WASMLD_ARGS=""
   CLANG_ARGS="-g"
 else
-  WASMLD_ARGS="-O2 --lto-O2 --strip-debug"
+  WASMLD_ARGS="-O2 --lto-O2 --strip-debug --gc-sections"
```
Comment (tracernz) [inline fbw-a32nx/src/wasm/fbw_a320/build.sh:16]: Suggest keeping this one in the common build flags so we strip unused cruft out of all binaries. Unfortunately GH won't let me suggest keeping the deleted line below (165).
```suggestion
  WASMLD_ARGS="-O2 --lto-O2 --strip-debug"
```

File: CMakeLists.txt [hunk 3]
```diff
@@ -10,7 +10,7 @@ set(FBW_COMMON ${FBW_ROOT}/fbw-common/src/wasm)
 include("${FBW_COMMON}/cpp-msfs-framework/cmake/TargetDefinition.cmake")
 
 # compiler refinement
-set(COMPILER_FLAGS "-Wall -Wextra -Wno-unused-function -Wno-unused-command-line-argument -Wno-ignored-attributes -Wno-macro-redefined -target wasm32-unknown-wasi --sysroot \"${MSFS_SDK}/WASM/wasi-sysroot\" -mthread-model single -fno-exceptions -fms-extensions -fvisibility=hidden -ffunction-sections -fdata-sections -fno-stack-protector")
+set(COMPILER_FLAGS "-Wall -Wextra -Wno-unused-function -Wno-unused-command-line-argument -Wno-ignored-attributes -Wno-macro-redefined -target wasm32-unknown-wasi --sysroot \"${MSFS_SDK}/WASM/wasi-sysroot\" -mthread-model single -fno-exceptions -fms-extensions -fvisibility=hidden -ffunction-sections -fwritable-strings -fdata-sections -Werror=return-type -fno-stack-protector -fstack-size-section -mbulk-memory")
```
Comment (tracernz) [inline CMakeLists.txt:13]: I don't think this flag is a good idea. It's to allow old code that expected to be able to write to string constants.
```suggestion
set(COMPILER_FLAGS "-Wall -Wextra -Wno-unused-function -Wno-unused-command-line-argument -Wno-ignored-attributes -Wno-macro-redefined -target wasm32-unknown-wasi --sysroot \"${MSFS_SDK}/WASM/wasi-sysroot\" -mthread-model single -fno-exceptions -fms-extensions -fvisibility=hidden -ffunction-sections -fdata-sections -Werror=return-type -fno-stack-protector -fstack-size-section -mbulk-memory")
```

File: fbw-a380x/src/wasm/fbw_a380/build.sh [hunk 4]
```diff
@@ -9,7 +9,7 @@ if [ "$1" == "--debug" ]; then
   WASMLD_ARGS=""
   CLANG_ARGS="-g"
 else
-  WASMLD_ARGS="-O2 --lto-O2 --strip-debug"
+  WASMLD_ARGS="-O2 --lto-O2 --strip-debug --gc-sections"
```
Comment (tracernz) [inline fbw-a380x/src/wasm/fbw_a380/build.sh:12]: As above, keep common.
```suggestion
  WASMLD_ARGS="-O2 --lto-O2 --strip-debug"
```

File: fbw-a380x/src/wasm/fbw_a380/build.sh [hunk 5]
```diff
@@ -40,6 +40,12 @@ clang \
   -fno-exceptions \
   -fms-extensions \
   -fvisibility=hidden \
+  -fdata-sections \
+  -fno-stack-protector \
+  -fstack-size-section \
+  -fwritable-strings \
```
Comment (tracernz) [inline fbw-a380x/src/wasm/fbw_a380/build.sh:46]:
```suggestion
```

File: fbw-common/src/wasm/cpp-msfs-framework/cmake/TargetDefinition.cmake [hunk 6]
```diff
@@ -23,7 +23,7 @@ macro(add_wasm_library)
         # we use `cp` to copy the unoptimized wasm file to the output directory
         set(WASM_OPT_FULL_CMD cp ${CMAKE_CURRENT_BINARY_DIR}/${ADD_WASM_LIBRARY_NAME}.wasm ${OUTPUT_DIRECTORY}/)
     else()
-        set(WASM_LD_ARGS -O2 --lto-O2 --strip-debug)
+        set(WASM_LD_ARGS -O2 --lto-O2 --strip-debug --gc-sections)
```
Comment (tracernz) [inline fbw-common/src/wasm/cpp-msfs-framework/cmake/TargetDefinition.cmake:26]:
```suggestion
        set(WASM_LD_ARGS -O2 --lto-O2 --strip-debug)
```

File: fbw-common/src/wasm/cpp-msfs-framework/cmake/TargetDefinition.cmake [hunk 7]
```diff
@@ -13,7 +13,7 @@ macro(add_wasm_library)
     endforeach()
 
     # wasm-ld general flags
-    set(CMAKE_WASM_LINKER_FLAGS --no-entry --allow-undefined --export __wasm_call_ctors --export-dynamic --export malloc --export free --export mallinfo --export mchunkit_begin --export mchunkit_next --export get_pages_state --export mark_decommit_pages --export-table --gc-sections -lc++ -lc++abi -L${MSFS_SDK}/WASM/wasi-sysroot/lib/wasm32-wasi -lc ${MSFS_SDK}/WASM/wasi-sysroot/lib/wasm32-wasi/libclang_rt.builtins-wasm32.a)
+    set(CMAKE_WASM_LINKER_FLAGS --no-entry --allow-undefined --export __wasm_call_ctors --export-dynamic --export malloc --export free --export mallinfo --export mchunkit_begin --export mchunkit_next --export get_pages_state --export mark_decommit_pages --export-table -lc++ -lc++abi -L${MSFS_SDK}/WASM/wasi-sysroot/lib/wasm32-wasi -lc ${MSFS_SDK}/WASM/wasi-sysroot/lib/wasm32-wasi/libclang_rt.builtins-wasm32.a)
```
Comment (tracernz) [inline fbw-common/src/wasm/cpp-msfs-framework/cmake/TargetDefinition.cmake:16]:
```suggestion
    set(CMAKE_WASM_LINKER_FLAGS --no-entry --allow-undefined --export __wasm_call_ctors --export-dynamic --export malloc --export free --export mallinfo --export mchunkit_begin --export mchunkit_next --export get_pages_state --export mark_decommit_pages --export-table --gc-sections -lc++ -lc++abi -L${MSFS_SDK}/WASM/wasi-sysroot/lib/wasm32-wasi -lc ${MSFS_SDK}/WASM/wasi-sysroot/lib/wasm32-wasi/libclang_rt.builtins-wasm32.a)
```

File: fbw-common/src/wasm/terronnd/build.sh [hunk 8]
```diff
@@ -8,7 +8,7 @@ OUTPUT="${DIR}/out/terronnd.wasm"
 if [ "$1" == "--debug" ]; then
   CLANG_ARGS="-g"
 else
-  WASMLD_ARGS="--strip-debug"
+  WASMLD_ARGS="--strip-debug --gc-sections"
```
Comment (tracernz) [inline fbw-common/src/wasm/terronnd/build.sh:11]: Keep the common one.
```suggestion
  WASMLD_ARGS="--strip-debug"
```

File: fbw-common/src/wasm/terronnd/build.sh [hunk 9]
```diff
@@ -47,6 +47,12 @@ clang++ \
   -fvisibility=hidden \
   -fno-common \
   -fstack-usage \
+  -fdata-sections \
+  -fno-stack-protector \
+  -fstack-size-section \
+  -fwritable-strings \
```
Comment (tracernz) [inline fbw-common/src/wasm/terronnd/build.sh:53]:
```suggestion
```

**Comment by Gurgel100** [comment]
Is it possible to upgrade directly to Rust 1.78?

**Comment by Saschl** [comment]
> Is it possible to upgrade directly to Rust 1.78?

We can try 👍 I'll update the rust version in this PR and if it works we can set it as default version in the dev env.

**Comment by Saschl** [comment]
Quality Assurance Report

Discord: saschl
Object of testing: https://github.com/flybywiresim/aircraft/pull/8585
Tier of Testing : 1
Date: 02/6/2024

Testing Process:

Full flight LOWI-EDSB.

Observed WASM based systems (FBW, Autothrottle, engines, electrics, hydraulics...). All working as intended.

Negatives:
N/A

Testing Results:
Passed

Conclusions:
LGTM

**Pull request merged**